File: src/chapters/minimization/content/code/bfgs.f90
program bfgs_demo
  implicit none
  integer, parameter :: n = 2
  real(8) :: x(n)
  x = bfgs([0d0, 0d0], 1d-8, 200)
  print '(A, 2F10.6)', 'x = ', x          ! -> 1, 2
contains
  function f(v) result(y)
    real(8), intent(in) :: v(n)
    real(8) :: y
    y = (v(1) - 1d0)**2 + 5d0*(v(2) - 2d0)**2
  end function f
  function grad(v) result(g)
    real(8), intent(in) :: v(n)
    real(8) :: g(n)
    g = [2d0*(v(1) - 1d0), 10d0*(v(2) - 2d0)]
  end function grad
  ! BFGS quasi-Newton minimization with backtracking (Armijo) line search.
  function bfgs(x0, tol, max_iter) result(x)
    real(8), intent(in) :: x0(n), tol
    integer, intent(in) :: max_iter
    real(8) :: x(n), g(n), d(n), s(n), y(n), g_new(n), x_new(n)
    real(8) :: Hm(n,n), Im(n,n), A(n,n), B(n,n), Hnew(n,n)
    real(8) :: t, fx, gd, sy, rho
    integer :: k, i, j
    x = x0
    Im = 0d0
    do i = 1, n
       Im(i,i) = 1d0
    end do
    Hm = Im                                  ! inverse-Hessian estimate
    g = grad(x)
    do k = 1, max_iter
       if (sqrt(sum(g**2)) < tol) exit
       d = -matmul(Hm, g)
       fx = f(x); gd = dot_product(g, d)
       t = 1d0
       do while (f(x + t*d) > fx + 1d-4*t*gd)
          t = t / 2d0
       end do
       s = t*d
       x_new = x + s
       g_new = grad(x_new)
       y = g_new - g
       sy = dot_product(s, y)
       if (sy > 1d-12) then                  ! H = (I - rho s y^T) H (I - rho y s^T) + rho s s^T
          rho = 1d0/sy
          do i = 1, n
             do j = 1, n
                A(i,j) = Im(i,j) - rho*s(i)*y(j)
                B(i,j) = Im(i,j) - rho*y(i)*s(j)
             end do
          end do
          Hnew = matmul(matmul(A, Hm), B)
          do i = 1, n
             do j = 1, n
                Hm(i,j) = Hnew(i,j) + rho*s(i)*s(j)
             end do
          end do
       end if
       x = x_new; g = g_new
    end do
  end function bfgs
end program bfgs_demo

File: src/chapters/minimization/content/code/bfgs.rs
fn nrm(v: &[f64]) -> f64 { v.iter().map(|x| x * x).sum::<f64>().sqrt() }
fn dot(a: &[f64], b: &[f64]) -> f64 { a.iter().zip(b).map(|(x, y)| x * y).sum() }

// BFGS quasi-Newton minimization with backtracking (Armijo) line search.
fn bfgs<F: Fn(&[f64]) -> f64, G: Fn(&[f64]) -> Vec<f64>>(
    f: F, grad: G, x0: &[f64], tol: f64, max_iter: usize,
) -> Vec<f64> {
    let n = x0.len();
    let mut x = x0.to_vec();
    let mut h = vec![vec![0.0; n]; n];           // inverse-Hessian estimate
    for i in 0..n { h[i][i] = 1.0; }
    let mut g = grad(&x);
    for _ in 0..max_iter {
        if nrm(&g) < tol { break; }
        let d: Vec<f64> = (0..n).map(|i| -dot(&h[i], &g)).collect();  // d = -H g
        let (fx, gd) = (f(&x), dot(&g, &d));
        let step = |t: f64| -> Vec<f64> { (0..n).map(|i| x[i] + t * d[i]).collect() };
        let mut t = 1.0;
        while f(&step(t)) > fx + 1e-4 * t * gd { t /= 2.0; }
        let s: Vec<f64> = d.iter().map(|di| t * di).collect();
        let x_new = step(t);
        let g_new = grad(&x_new);
        let y: Vec<f64> = (0..n).map(|i| g_new[i] - g[i]).collect();
        let sy = dot(&s, &y);
        if sy > 1e-12 {                          // H = (I - rho s y^T) H (I - rho y s^T) + rho s s^T
            let rho = 1.0 / sy;
            let a: Vec<Vec<f64>> = (0..n).map(|i| (0..n).map(|j|
                (if i == j { 1.0 } else { 0.0 }) - rho * s[i] * y[j]).collect()).collect();
            let b: Vec<Vec<f64>> = (0..n).map(|i| (0..n).map(|j|
                (if i == j { 1.0 } else { 0.0 }) - rho * y[i] * s[j]).collect()).collect();
            let ah: Vec<Vec<f64>> = (0..n).map(|i| (0..n).map(|j|
                (0..n).map(|l| a[i][l] * h[l][j]).sum()).collect()).collect();
            let ahb: Vec<Vec<f64>> = (0..n).map(|i| (0..n).map(|j|
                (0..n).map(|l| ah[i][l] * b[l][j]).sum()).collect()).collect();
            for i in 0..n { for j in 0..n { h[i][j] = ahb[i][j] + rho * s[i] * s[j]; } }
        }
        x = x_new; g = g_new;
    }
    x
}

fn main() {
    let f = |v: &[f64]| (v[0] - 1.0).powi(2) + 5.0 * (v[1] - 2.0).powi(2);
    let grad = |v: &[f64]| vec![2.0 * (v[0] - 1.0), 10.0 * (v[1] - 2.0)];
    println!("{:?}", bfgs(f, grad, &[0.0, 0.0], 1e-8, 200));   // -> [1, 2]
}
